File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, h = nums.size() - 1;
        while (l <= h){
            int m = (l + h) >> 1;
            if (nums[m] == target) return m;
            //LEFT HALF SORTED
            if (nums[l] <= nums[m]){
                if(nums[l] <= target && nums[m] >= target) {
                    h = m - 1;
                } else l = m + 1;
            } else {
                if(nums[m] <= target && nums[h] >= target){
                    l = m + 1;
                } else {
                    h = m - 1;
                }
            }
        }
        return -1;
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, A: List[int], target: int) -> int:
        l, h = 0 , len(A) - 1
        while l <= h:
            m = (l + h) >> 1
            if A[m] == target: return m
            elif A[l] <= A[m]:
                if A[l] <= target and target <= A[m]:
                    h = m - 1
                else:
                    l = m + 1
            else:
                if A[m] <= target and target <= A[h]:
                    l = m + 1
                else:
                    h = m - 1
                
        return -1